File: Module/PlayWays/src/main/java/com/module/playways/BaseRoomData.java
package com.module.playways;

import com.common.core.userinfo.model.UserInfoModel;
import com.module.playways.rank.prepare.model.PlayerInfoModel;
import com.module.playways.rank.prepare.model.BaseRoundInfoModel;
import com.module.playways.rank.song.model.SongModel;

import java.io.Serializable;
import java.util.List;

import io.reactivex.subjects.PublishSubject;


/**
 * 房间内所有数据的聚合类
 * 每种模式的房间内状态信息都由其存储
 */
public abstract class BaseRoomData<T extends BaseRoundInfoModel> implements Serializable {
    public final static String TAG = "RoomData";

    public final static String RANK_BATTLE_START_SVGA = "http://res-static.inframe.mobi/app/rank_battle_start.svga";
    public final static String RANK_RESULT_WIN_SVGA = "http://res-static.inframe.mobi/app/rank_result_win.svga";
    public final static String RANK_RESULT_LOSE_SVGA = "http://res-static.inframe.mobi/app/rank_result_lose.svga";
    public final static String RANK_RESULT_DRAW_SVGA = "http://res-static.inframe.mobi/app/rank_result_draw.svga";
    public final static String GRAB_BURST_BIG_SVGA = "http://res-static.inframe.mobi/app/grab_burst_big_animation.svga";

    public final static String PK_MAIN_STAGE_WEBP = "http://res-static.inframe.mobi/app/pk_main_stage.webp";
    public final static String READY_GO_SVGA_URL = "http://res-static.inframe.mobi/app/sige_go.svga";
    public final static String ROOM_STAGE_SVGA = "http://res-static.inframe.mobi/app/main_stage_people.svga";
    public final static String ROOM_SPECAIL_EMOJI_DABIAN = "http://res-static.inframe.mobi/app/emoji_bianbian.svga";
    public final static String ROOM_SPECAIL_EMOJI_AIXIN = "http://res-static.inframe.mobi/app/emoji_love.svga";
    public static final String AUDIO_FOR_AI_PATH = "audioforai.aac";
    public static final String MATCHING_SCORE_FOR_AI_PATH = "matchingscore.json";

    protected int mGameId; // 房间id

    protected String mSysAvatar; // 系统头像

    /**
     * 当要拿服务器时间和本地时间比较时，请将服务器时间加上这个矫正值
     * 如
     * if(System.currentTimeMillis() > mGameStartTs + mShiftts){
     * <p>
     * }
     */
    protected int mShiftTs;// 本地时间比服务器快多少毫秒，比如快1秒，mShiftTs = 1000;

    protected long mGameCreateTs;// 游戏创建时间,服务器的

    protected long mGameStartTs;// 游戏开始时间,服务器的

    protected long mGameOverTs;// 游戏结束时间,服务器的

    protected long mLastSyncTs;// 上次同步服务器状态时间,服务器的

    protected SongModel mSongModel; // 歌曲信息

    protected T mExpectRoundInfo;// 按理的 期望的当前的轮次

    protected T mRealRoundInfo;// 实际的当前轮次信息

    protected boolean mIsGameFinish = false; // 游戏开始了

    protected boolean mMute = false;//是否mute

    private String mAgoraToken; // 声网token

    public abstract int getGameType();

    public abstract void checkRoundInEachMode();

    public void setIsGameFinish(boolean isGameFinish) {
        this.mIsGameFinish = isGameFinish;
    }

    public boolean isIsGameFinish() {
        return mIsGameFinish;
    }

    public int getGameId() {
        return mGameId;
    }

    public void setGameId(int gameId) {
        mGameId = gameId;
    }

    public String getSysAvatar() {
        return mSysAvatar;
    }

    public void setSysAvatar(String sysAvatar) {
        mSysAvatar = sysAvatar;
    }

    public int getShiftTs() {
        return mShiftTs;
    }

    public void setShiftTs(int shiftTs) {
        mShiftTs = shiftTs;
    }

    public long getGameCreateTs() {
        return mGameCreateTs;
    }

    public void setGameCreateTs(long gameCreateTs) {
        mGameCreateTs = gameCreateTs;
    }

    public long getGameStartTs() {
        return mGameStartTs;
    }

    public void setGameStartTs(long gameStartTs) {
        mGameStartTs = gameStartTs;
    }

    public long getGameOverTs() {
        return mGameOverTs;
    }

    public void setGameOverTs(long gameOverTs) {
        mGameOverTs = gameOverTs;
    }

    public long getLastSyncTs() {
        return mLastSyncTs;
    }

    public void setLastSyncTs(long lastSyncTs) {
        mLastSyncTs = lastSyncTs;
    }

    public SongModel getSongModel() {
        return mSongModel;
    }

    public void setSongModel(SongModel songModel) {
        mSongModel = songModel;
    }

    public T getExpectRoundInfo() {
        return mExpectRoundInfo;
    }

    public void setExpectRoundInfo(T expectRoundInfo) {
        mExpectRoundInfo = expectRoundInfo;
    }

    public <T extends BaseRoundInfoModel> T getRealRoundInfo() {
        return (T) mRealRoundInfo;
    }

    public void setRealRoundInfo(T realRoundInfo) {
        mRealRoundInfo = realRoundInfo;
    }


    public int getRealRoundSeq() {
        if (mRealRoundInfo != null) {
            return mRealRoundInfo.getRoundSeq();
        }

        return -1;
    }

    public boolean isMute() {
        return mMute;
    }

    public void setMute(boolean mute) {
        mMute = mute;
    }

    @Override
    public String toString() {
        return "RoomData{" +
                "mGameType=" + getGameType() +
                ", mGameId=" + mGameId +
                ", mSysAvatar='" + mSysAvatar + '\'' +
                ", mShiftTs=" + mShiftTs +
                ", mGameCreateTs=" + mGameCreateTs +
                ", mGameStartTs=" + mGameStartTs +
                ", mGameOverTs=" + mGameOverTs +
                ", mLastSyncTs=" + mLastSyncTs +
                ", mSongModel=" + mSongModel +
                ", mExpectRoundInfo=" + mExpectRoundInfo +
                ", mRealRoundInfo=" + mRealRoundInfo +
                ", mIsGameFinish=" + mIsGameFinish +
                ", mMute=" + mMute +
                '}';
    }

    public abstract <T extends PlayerInfoModel> List<T> getPlayerInfoList();

    public UserInfoModel getUserInfo(int userID) {
        if (userID == 0) {
            return null;
        }
        List<PlayerInfoModel> l = getPlayerInfoList();
        if (l == null) {
            return null;
        }
        for (PlayerInfoModel playerInfo : l) {
            if (playerInfo.getUserInfo().getUserId() == userID) {
                return playerInfo.getUserInfo();
            }
        }
        return null;
    }

    public void setAgoraToken(String agoraToken) {
        mAgoraToken = agoraToken;
    }

    public String getAgoraToken() {
        return mAgoraToken;
    }
}

File: Module/PlayWays/src/main/java/com/module/playways/BaseRoomData.kt
package com.module.playways

import android.util.LruCache
import com.common.core.userinfo.model.UserInfoModel
import com.module.playways.grab.room.event.GrabMyCoinChangeEvent
import com.module.playways.room.gift.event.UpdateHZEvent
import com.module.playways.room.prepare.model.BaseRoundInfoModel
import com.module.playways.room.prepare.model.PlayerInfoModel
import com.module.playways.room.song.model.SongModel
import org.greenrobot.eventbus.EventBus
import java.io.Serializable


/**
 * 房间内所有数据的聚合类
 * 每种模式的房间内状态信息都由其存储
 */
abstract class BaseRoomData<T : BaseRoundInfoModel> : Serializable {
    val TAG = "RoomData"

    var gameId: Int = 0 // 房间id

    var sysAvatar: String? = null // 系统头像

    /**
     * 当要拿服务器时间和本地时间比较时，请将服务器时间加上这个矫正值
     * 如
     * if(System.currentTimeMillis() > mGameStartTs + mShiftts){
     *
     *
     * }
     */
    var shiftTs: Int = 0// 本地时间比服务器快多少毫秒，比如快1秒，mShiftTs = 1000;

    var gameCreateTs: Long = 0// 游戏创建时间,服务器的,绝对时间

    var gameStartTs: Long = 0// 游戏开始时间,服务器的，相对于 gameCreateTs 的相对时间

    var gameOverTs: Long = 0// 游戏结束时间,服务器的

    var lastSyncTs: Long = 0// 上次同步服务器状态时间,服务器的

    var songModel: SongModel? = null // 歌曲信息

    var expectRoundInfo: T? = null// 按理的 期望的当前的轮次

    var realRoundInfo: T? = null// 实际的当前轮次信息

    var isIsGameFinish = false // 游戏开始了

    var isMute = false//是否mute

    var agoraToken: String? = null // 声网token

    abstract val gameType: Int

    private var lastHzTs: Long = -1 // 红钻更新的时间戳

    protected var mCoin: Int = 0// 金币数
    protected var mHzCount: Float = 0.toFloat()// 金币数

    val realRoundSeq: Int
        get() = if (realRoundInfo != null) {
            realRoundInfo!!.roundSeq
        } else -1

    abstract fun checkRoundInEachMode()

    fun getCoin(): Int {
        return mCoin
    }

    fun setCoin(coin: Int) {
        if (this.mCoin != coin) {
            EventBus.getDefault().post(GrabMyCoinChangeEvent(coin, coin - this.mCoin))
            this.mCoin = coin
        }
    }

    fun setCoinNoEvent(coin: Int) {
        if (this.mCoin != coin) {
            this.mCoin = coin
        }
    }

    fun setHzCount(hzCount: Float, ts: Long) {
        if (lastHzTs < ts) {
            lastHzTs = ts
            mHzCount = hzCount
            UpdateHZEvent.sendEvent(hzCount, ts)
        }
    }

    fun getHzCount(): Float {
        return mHzCount
    }

    override fun toString(): String {
        return "RoomData{" +
                "mGameType=" + gameType +
                ", gameId=" + gameId +
                ", mSysAvatar='" + sysAvatar + '\''.toString() +
                ", mShiftTs=" + shiftTs +
                ", mGameCreateTs=" + gameCreateTs +
                ", mGameStartTs=" + gameStartTs +
                ", mGameOverTs=" + gameOverTs +
                ", mLastSyncTs=" + lastSyncTs +
                ", mSongModel=" + songModel +
                ", mExpectRoundInfo=" + expectRoundInfo +
                ", mRealRoundInfo=" + realRoundInfo +
                ", mIsGameFinish=" + isIsGameFinish +
                ", mMute=" + isMute +
                '}'.toString()
    }

    abstract fun <T : PlayerInfoModel> getPlayerInfoList(): List<T>?

    @Transient
    private val userInfoMap = LruCache<Int, PlayerInfoModel>(20)

    fun getUserInfo(userID: Int?): UserInfoModel? {
        if (userID == null || userID == 0) {
            return null
        }
        val playerInfoModel = userInfoMap[userID]
        if (playerInfoModel == null) {
            val l = getPlayerInfoList<PlayerInfoModel>() ?: return null
            for (playerInfo in l) {
                if (playerInfo.userInfo.userId == userID) {
                    userInfoMap.put(playerInfo.userInfo.userId, playerInfo)
                    return playerInfo.userInfo
                }
            }
        } else {
            return playerInfoModel.userInfo
        }
        return null
    }

    fun <T : PlayerInfoModel> getPlayerInfoModel(userID: Int?): T? {
        if (userID == null || userID == 0) {
            return null
        }
        val playerInfoModel = userInfoMap[userID]
        if (playerInfoModel == null) {
            val l = getPlayerInfoList<PlayerInfoModel>() ?: return null
            for (playerInfo in l) {
                if (playerInfo.userInfo.userId == userID) {
                    userInfoMap.put(playerInfo.userInfo.userId, playerInfo)
                    return playerInfo as T?
                }
            }
        } else {
            return playerInfoModel as T?
        }
        return null
    }

    open fun <User : PlayerInfoModel> getInSeatPlayerInfoList(): List<User>? {
        return ArrayList()
    }

    companion object {

        val RANK_BATTLE_START_SVGA = "http://res-static.inframe.mobi/app/rank_battle_start.svga"
        val RANK_RESULT_WIN_SVGA = "http://res-static.inframe.mobi/app/rank_result_win.svga"
        val RANK_RESULT_LOSE_SVGA = "http://res-static.inframe.mobi/app/rank_result_lose.svga"
        val RANK_RESULT_DRAW_SVGA = "http://res-static.inframe.mobi/app/rank_result_draw.svga"
        val GRAB_BURST_BIG_SVGA = "http://res-static.inframe.mobi/app/grab_burst_big_animation.svga"

        val PK_MAIN_STAGE_WEBP = "http://res-static.inframe.mobi/app/pk_main_stage.webp"
        val READY_GO_SVGA_URL = "http://res-static.inframe.mobi/app/sige_go.svga"
        val ROOM_STAGE_SVGA = "http://res-static.inframe.mobi/app/main_stage_people.svga"
        val ROOM_SPECAIL_EMOJI_DABIAN = "http://res-static.inframe.mobi/app/emoji_bianbian.svga"
        val ROOM_SPECAIL_EMOJI_AIXIN = "http://res-static.inframe.mobi/app/emoji_love.svga"
        val AUDIO_FOR_AI_PATH = "audioforai.aac"
        val MATCHING_SCORE_FOR_AI_PATH = "matchingscore.json"
    }

}
